### ISSUE TITLE
Advice needed:  opencv-rust won't build on linux with vcpkg and newer versions of opencv

### ISSUE THREAD

**Issue by jaysonsantos** [twistedfall/opencv-rust#167]
Hey there first of all thanks for these bindings they are really helpful :).
I am trying to use newer OpenCV versions and using vcpkg to install it on Linux, but there is a linking problem when putting the final binary together, could you take a look and see if you think I did something wrong? I tried to follow the CI stuff you have here but mixing Linux and vcpkg.
I will also try to compile the instructions in a doc to put here for future reference.
To make it easier to reproduce, I put it all in the following Dockerfile:

```dockerfile
FROM rust:1.46

WORKDIR /app

RUN apt-get update && \
    DEBIAN_FRONTEND=noninteractive apt-get -y install libclang-dev clang curl git zip unzip pkg-config

ENV VCPKG_ROOT /usr/src/vcpkg
ENV VCPKGRS_DYNAMIC=1
ENV PATH "/bin:/sbin:/usr/bin:/usr/local/bin:/usr/sbin:/root/.cargo/bin:/usr/local/cargo/bin"

RUN bash -exc 'cd /usr/src && git clone https://github.com/microsoft/vcpkg && cd vcpkg && ./bootstrap-vcpkg.sh && ./vcpkg install opencv4[contrib,nonfree] && ./vcpkg list'

RUN USER=test cargo new test-app && cd test-app && \
    echo 'opencv = { version = "0.46", default-features = false, features = ["contrib", "buildtime-bindgen", "opencv-4", "clang-runtime"] }' >> Cargo.toml && \
    echo 'use opencv::videoio;fn main() {let _cam = videoio::VideoCapture::new(0, videoio::CAP_ANY).unwrap();}' > src/main.rs && \
    cargo build -vvv
```

Here you can see the build log https://gist.github.com/jaysonsantos/6f0042734ea754045e28ad57d01409ba

**Comment by jaysonsantos** [comment]
Sorry, I said before it was linking problems but I had them on windows but on linux it seems like headers problem

**Comment by twistedfall** [comment]
Thanks for the report, I will check it out! So far I haven't tried running vcpkg on Linux, so this scenario is untested, but I'll try to fix it as soon as possible.

**Comment by jaysonsantos** [comment]
Hey there, I will also try and figure out how it works.
I am quite a newbie in c/c++ but I will give it a try.
I saw similar errors on the internet but when linking, and some people said that the ordering which you link for example `opencv_core` actually matters on that.

**Comment by twistedfall** [comment]
The problem is in the linker argument order as crate receives it from vcpkg. I’ll need some time to think about how to work it around more gracefully.

**Comment by jaysonsantos** [comment]
vcpkg reads from `$VCPKG_ROOT/installed/vcpkg/status` which does not necessarily have the right order.
I guess it should probably better to detect that `VCPKG_ROOT` and then call cmake  with `-DCMAKE_TOOLCHAIN_FILE=$VCPKG_ROOT/scripts/buildsystems/vcpkg.cmake` what do you think?

**Comment by jaysonsantos** [comment]
This changes makes it work:
```diff
diff --git a/build.rs b/build.rs
index ae0302d..2b019c0 100644
--- a/build.rs
+++ b/build.rs
@@ -610,6 +610,19 @@ impl Library {
                eprintln!("=== Probing OpenCV library using cmake");
                let cmake_pkg = PackageName::cmake();
                let cmake_bin = env::var_os("OPENCV_CMAKE_BIN").unwrap_or_else(|| "cmake".into());
+               let vcpkg_toolchain_file = env::var_os("VCPKG_ROOT")
+                       .and_then(|root| Some(format!("{}/scripts/buildsystems/vcpkg.cmake", root.to_string_lossy())))
+                       .unwrap_or_else(|| "".to_string());
+               let get_args = |mode| {
+                       vec![
+                               "--find-package".into(),
+                               "-DCMAKE_TOOLCHAIN_FILE=/home/jayson/src/vcpkg/scripts/buildsystems/vcpkg.cmake".into(),
+                               "-DCOMPILER_ID=GNU".into(),
+                               "-DLANGUAGE=CXX".into(),
+                               format!("-DMODE={}", mode),
+                               vcpkg_toolchain_file.clone(),
+                       ]
+               };
 
                let include_paths = include_paths
                        .map(|paths| Self::list(paths)
@@ -619,12 +632,7 @@ impl Library {
                        .ok_or_else(|| "Nobody is going to see that")
                        .or_else(|_| Command::new(&cmake_bin)
                                .current_dir(&*OUT_DIR)
-                               .args(&[
-                                       "--find-package",
-                                       "-DCOMPILER_ID=GNU",
-                                       "-DLANGUAGE=CXX",
-                                       "-DMODE=COMPILE",
-                               ])
+                               .args(&get_args("COMPILE"))
                                .arg(format!("-DNAME={}", cmake_pkg))
                                .output()
                                .map_err(Box::<dyn std::error::Error>::from)
@@ -661,12 +669,7 @@ impl Library {
                        if link_paths.is_none() || link_libs.is_none() {
                                Command::new(&cmake_bin)
                                        .current_dir(&*OUT_DIR)
-                                       .args(&[
-                                               "--find-package",
-                                               "-DCOMPILER_ID=GNU",
-                                               "-DLANGUAGE=CXX",
-                                               "-DMODE=LINK",
-                                       ])
+                                       .args(&get_args("LINK"))
                                        .arg(format!("-DNAME={}", cmake_pkg))
                                        .output()
                                        .map_err(Box::<dyn std::error::Error>::from)
```
But I had to change OpenCV's cmake file with `sed -i.bkp 's/find_dependency.Threads.//' $VCPKG_ROOT/installed/x64-linux/share/opencv/OpenCVModules.cmake`

**Comment by twistedfall** [comment]
Thank you for the patch, it's really helpful! Can you please share the details on why you needed to modify the OpenCV's cmake file?

**Comment by jaysonsantos** [comment]
Hey there, that makefile make a call to `find_dependency(Threads)` otherwise it would break with this:

```
=== Can't probe using cmake, continuing with other methods, error: cmake returned an error
    stdout: "OpenCV not found.\n"
    stderr: "CMake Error at /usr/share/cmake-3.16/Modules/FindThreads.cmake:49 (message):\n  FindThreads only works if either C or CXX language is enabled\nCall Stack (most recent call first):\n  /home/jayson/src/vcpkg/scripts/buildsystems/vcpkg.cmake:493 (_find_package)\n  /usr/share/cmake-3.16/Modules/CMakeFindDependencyMacro.cmake:47 (find_package)\n  /home/jayson/src/vcpkg/installed/x64-linux/share/opencv/OpenCVModules.cmake:31 (find_dependency)\n  /home/jayson/src/vcpkg/installed/x64-linux/share/opencv/OpenCVConfig.cmake:126 (include)\n  /home/jayson/src/vcpkg/scripts/buildsystems/vcpkg.cmake:493 (_find_package)\n  /usr/share/cmake-3.16/Modules/CMakeFindPackageMode.cmake:183 (find_package)\n\n\nCMake Error: Problem processing arguments. Aborting.\n\n"
=
```

the area where it fails on FindTHreads.cmake is this
```cmake
if(CMAKE_C_COMPILER_LOADED)
  include (CheckIncludeFile)
  include (CheckCSourceCompiles)
elseif(CMAKE_CXX_COMPILER_LOADED)
  include (CheckIncludeFileCXX)
  include (CheckCXXSourceCompiles)
else()
  message(FATAL_ERROR "FindThreads only works if either C or CXX language is enabled")
endif()
```

I assume it only works when cmake is being called to actually compile

**Comment by twistedfall** [comment]
Hmm, looks like `--find-package` is actually deprecated: https://gitlab.kitware.com/cmake/cmake/-/issues/20849 I'm in the process of rewriting the cmake probe code, it should help with that problem.

**Comment by jaysonsantos** [comment]
What do you think of build.rs building this to get the right values?

CMakeLists.txt
```cmake
cmake_minimum_required(VERSION 3.16)

set(CMAKE_TOOLCHAIN_FILE "$ENV{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake")

project(opencv_flags)

set(CMAKE_CXX_STANDARD 14)

find_package(OpenCV REQUIRED)

list(JOIN OpenCV_INCLUDE_DIRS "," JOINED_INCLUDE_DIRS)
list(JOIN OpenCV_LIBS "," JOINED_LIBS)

add_compile_definitions(OpenCV_INCLUDE_DIRS="${JOINED_INCLUDE_DIRS}")
add_compile_definitions(OpenCV_LIBS="${JOINED_LIBS}")

add_executable(opencv_flags main.cpp)
```

main.cpp
```c++
#include <iostream>

int main() {
    std::cout << OpenCV_INCLUDE_DIRS << std::endl;
    std::cout << OpenCV_LIBS << std::endl;
    return 0;
}
```

**Comment by twistedfall** [comment]
I came up with the following:
```cmake
project(OpenCVProbe)

cmake_minimum_required(VERSION 3.0)

find_package(${OCVRS_PACKAGE_NAME} CONFIG REQUIRED)

message("OCVRS_INCLUDE_DIRS:${OpenCV_INCLUDE_DIRS}")
message("OCVRS_VERSION:${OpenCV_VERSION}")

foreach(OCV_LIB ${OpenCV_LIBS})
	find_library(OCV_LIB_${OCV_LIB}_PATH "${OCV_LIB}" HINTS "${OpenCV_INSTALL_PATH}/lib" REQUIRED)
	message("OCVRS_LIB:${OCV_LIB}:${OCV_LIB_${OCV_LIB}_PATH}")
endforeach()
```

This is for generic (not vcpkg specific) cmake probing, It communicates those values over the stderr. It was my first time writing a cmake file though, what do you think?

**Comment by jaysonsantos** [comment]
Hey nice, you don't even need the actual code?
To support both vcpkg and plain cmake this could do the trick:

```cmake
if(DEFINED ENV{VCPKG_ROOT})
set(CMAKE_TOOLCHAIN_FILE "$ENV{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake")
endif()

project(OpenCVProbe)

cmake_minimum_required(VERSION 3.0)

find_package(${OCVRS_PACKAGE_NAME} CONFIG REQUIRED)

message("OCVRS_INCLUDE_DIRS:${OpenCV_INCLUDE_DIRS}")
message("OCVRS_VERSION:${OpenCV_VERSION}")

foreach(OCV_LIB ${OpenCV_LIBS})
	find_library(OCV_LIB_${OCV_LIB}_PATH "${OCV_LIB}" HINTS "${OpenCV_INSTALL_PATH}/lib" REQUIRED)
	message("OCVRS_LIB:${OCV_LIB}:${OCV_LIB_${OCV_LIB}_PATH}")
endforeach()
```
With some tests, I saw that the toolchain file has to be before the project

**Comment by twistedfall** [comment]
Can you please test the `cmake-rework` branch? It should just work without any modification of the vcpkg installed cmake files.

**Comment by jaysonsantos** [comment]
Hey there, for the linux image it seems to be compiling fine.
I think that only on windows the linker has some problems.
I will try to make it reproducible and open a new issue.
Thank you!

**Comment by twistedfall** [comment]
The fix is now released in v0.46.1
